lib/remote/filterutility.cpp

```diff
@@ -178,7 +178,7 @@ bool FilterUtility::HasPermission(const ApiUser::Ptr& user, const String& permis
 				if (!*permissionFilter)
 					*permissionFilter = fexpr;
 				else
-					*permissionFilter = new LogicalOrExpression(std::unique_ptr<Expression>(*permissionFilter), std::unique_ptr<Expression>(fexpr));
+					*permissionFilter = new LogicalOrExpression(std::unique_ptr<Expression>(permissionFilter->get()), std::unique_ptr<Expression>(fexpr));
```

Comment:
# Riddle

Why will this crash?

Reply:
I don't see any reason to crash here 💥 but if you do, I'd be eager to get it from you.

Consider the following example:
```cxx
std::unique_ptr<string> foo = std::make_unique<string>("Foo Bar");;
std::unique_ptr<string>* fooPtr = &foo;

cout << foo.get() << "\n";
cout << &foo << "\n";

cout << fooPtr->get() << "\n";
cout << &*fooPtr << "\n";
cout << &fooPtr << "\n";

cout << *fooPtr->get() << "\n";
cout << *foo << "\n";
 
*fooPtr = std::make_unique<string>("Foo Bar Foo: Previous Ptr: " + *fooPtr->get());
cout << *fooPtr->get() << "\n";
cout << *foo << "\n";
```

The output from this example looks like this, so it's pretty much the same thing what we' re doing with the `permissionFilter`.

```cxx
0x55e7c24a9eb0                        // cout << foo.get() << "\n";
0x7fffc07eaaf0                        // cout << &foo << "\n";
0x55e7c24a9eb0                        // cout << fooPtr->get() << "\n";
0x7fffc07eaaf0                        // cout << &*fooPtr << "\n";
0x7fffc07eaaf8                        // cout << &fooPtr << "\n";

Foo Bar                               // cout << *fooPtr->get() << "\n";
Foo Bar                               // cout << *foo << "\n";

// *fooPtr = std::make_unique<string>("Foo Bar Foo: Previous Ptr: " + *fooPtr->get());
Foo Bar Foo: Previous Ptr: Foo Bar    // cout << *fooPtr->get() << "\n";
Foo Bar Foo: Previous Ptr: Foo Bar    // cout << *foo << "\n";
```

Reply:
## Reason

```
object ApiUser "crashy" {
  password = "crashy"
  permissions = [
    { "permission" = "objects/query/Host"; "filter" = {{ true }} },
    { "permission" = "objects/query/Host"; "filter" = {{ true }} }
  ]
}
```

Reply:
```
$ curl -fksSiLu crashy:crashy -H 'Accept: application/json' 'https://127.0.0.1:5665/v1/objects/hosts'
curl: (52) Empty reply from server
$ curl -fksSiLu crashy:crashy -H 'Accept: application/json' 'https://127.0.0.1:5665/v1/objects/hosts'
curl: (7) Failed to connect to 127.0.0.1 port 5665 after 6 ms: Connection refused
$
```

Reply:
Overseen the 🐘:

```cxx
// This sets ref count of the permission filter to `1` and ...
if (!permissionFilter->get())
       *permissionFilter = fexpr;
else
   // when calling `permissionFilter->get()` and reassign it with a new value, it will call the destructor and as the
   // previous ref count is still `1`, it will destroy the former value of `permissionFilter` which is `FunctionCallExpression*`
    *permissionFilter = new LogicalOrExpression(std::unique_ptr<Expression>(permissionFilter->get()), std::unique_ptr<Expression>(fexpr));
```

**Hack:**
Use `permissionFilter->detach()` instead of `permissionFilter->get()`

Reply:
I'd really try to avoid converting from `X::Ptr` to `unique_ptr<T>` because this is only valid if you can ensure that this is the only `X::Ptr` to that object. This function now only uses `detach()` because it also uses the output parameter as an intermediate variable where it passes the ownership of the `Expression` object to `intrusive_ptr`, only to later say "oops, didn't actually mean to do this". Would be better to use a local variable of type `unique_ptr<Expression>` which can then just be moved into the `LogicalOrExpression` and at the end, transfer ownership from `unique_ptr` to `intrusive_ptr` once (which is guaranteed to be safe as `unique_ptr` has exclusive ownership of the object).